spec/bson/dbref_spec.rb

```diff
@@ -202,17 +230,95 @@
       it 'sets the database to nil' do
         expect(decoded.database).to be_nil
       end
+
+      it 'is of class DBRef' do
+        expect(decoded).to be_a described_class
+      end
     end
 
     context 'when other keys exist' do
 
-      let(:dbref) do
-        described_class.new({ '$ref' => 'users', '$id' => object_id, 'x' => 'y' })
+      let(:hash) do
+        { '$ref' => 'users', '$id' => object_id, 'x' => 'y' }
       end
 
       it 'decodes the key' do
         expect(decoded['x']).to eq('y')
       end
+
+      it 'is of class DBRef' do
+        expect(decoded).to be_a described_class
+      end
+    end
+
+    context 'when nesting the dbref' do
+      context 'when it is a valid dbref' do
+        let(:hash) do
+          { 'dbref' => { '$ref' => 'users', '$id' => object_id } }
+        end
+
+        it 'should not raise' do
+          expect do
+            buffer
+          end.to_not raise_error
+        end
+
+        it 'has the correct class' do
+          expect(decoded['dbref']).to be_a described_class
+        end
+      end
+
+      context 'when it is an invalid dbref' do
+        let(:hash) do
+          { 'dbref' => { '$ref' => 1, '$id' => object_id } }
+        end
+
+        it 'should not raise' do
+          expect do
+            decoded
+          end.to_not raise_error
+        end
+
+        it 'has the correct class' do
+          expect(decoded['dbref']).to be_a BSON::Document
+        end
+      end
+    end
+
+    context 'when nesting a dbref inside a dbref' do
+      context 'when it is a valid dbref' do
+        let(:hash) do
+          { 'dbref' => { '$ref' => 'users', '$id' => object_id, 'dbref' => { '$ref' => 'users', '$id' => object_id } } }
+        end
+
+        it 'should not raise' do
+          expect do
+            buffer
+          end.to_not raise_error
+        end
+
+        it 'has the correct class' do
+          expect(decoded['dbref']).to be_a described_class
+          expect(decoded['dbref']['dbref']).to be_a described_class
```

Comment:
This fails, because it is actually of type `BSON::Document`.

The reason this happens is because in `from_bson` we call the `BSON::DBRef` constructor, which just validates the dbref and calls the `BSON::Document` constructor (super). The DBRef constructor does not further parse for more DBRefs (and I'm not sure it should?)